### PR TITLE
bugfix: do not lose file-upload after validation error

### LINKED ISSUE #46: Broken image saved on form validation error
There seems to be a problem with uploading images in "z3c.form" forms. If you have a dexterity content type with an image field that uses the `INamedImageWidget` widget and there was a form validation error, the image will not save correctly.

**Steps to reproduce:**
1. Set one field of the `Image` dexterity content type as required (e.g. `title`, https://demo-latest-py3.plone.org/dexterity-types/Image/@@fields)
1. Try to add a new `Image` (https://demo-latest-py3.plone.org/de/++add++Image)
1. Click `Choose file` and select a image file from your Computer.
1. Leave the required field blank (e.g. no `title`)
1. Click `Save`
1. --> Form validation kicks in and after reloading the status message "Error There were some errors." will be shown.
1. Now correct your error and fill out the required field
2. Make sure to choose `Keep existing image` for your image field
1. Click `Save` again

**Screenshots:**
   

**Expected:**
A new `Image` dexterity content with my uploaded file should show up.

**Actual:**
A new `Image` dexterity content was created, but without any image file. The image field seems to be empty. If you try to view the image (..`/image/view`), you will get the following error

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 162, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 371, in publish_module
  Module ZPublisher.WSGIPublisher, line 266, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module zope.browserpage.simpleviewclass, line 41, in __call__
  Module Products.Five.browser.pagetemplatefile, line 126, in __call__
  Module Products.Five.browser.pagetemplatefile, line 58, in __call__
  Module zope.pagetemplate.pagetemplate, line 133, in pt_render
  Module Products.PageTemplates.engine, line 378, in __call__
  Module z3c.pt.pagetemplate, line 176, in render
  Module chameleon.zpt.template, line 307, in render
  Module chameleon.template, line 214, in render
  Module chameleon.utils, line 75, in raise_with_traceback
  Module chameleon.template, line 192, in render
  Module bad2e2a69ccbaee8d9cbacb9808051f9, line 501, in render
  Module feae8a7feccd5d46d0ebdc5ee7663dfe, line 860, in render_master
  Module feae8a7feccd5d46d0ebdc5ee7663dfe, line 1489, in render_content
  Module bad2e2a69ccbaee8d9cbacb9808051f9, line 486, in __fill_content_core
  Module bad2e2a69ccbaee8d9cbacb9808051f9, line 115, in render_content_core
  Module zope.tales.expressions, line 250, in __call__
  Module Products.PageTemplates.Expressions, line 188, in _eval
  Module zope.tales.expressions, line 153, in _eval
  Module Products.PageTemplates.Expressions, line 104, in trustedBoboAwareZopeTraverse
  Module zope.traversing.adapters, line 156, in traversePathElement
   - __traceback_info__: (None, 'getSize')
  Module zope.traversing.adapters, line 61, in traverse
   - __traceback_info__: (None, 'getSize', [])
zope.location.interfaces.LocationError: zope.location.interfaces.LocationError: (None, 'getSize')

 - Expression: "context/image/getSize"
 - Filename:   ... .8.egg/plone/app/contenttypes/browser/templates/image.pt
 - Location:   (line 11: col 58)
 - Source:     ... core" tal:define="size context/image/getSize;
                                          ^^^^^^^^^^^^^^^^^^^^^
 - Expression: "provider:plone.abovecontentbody"
 - Filename:   ... one/Products/CMFPlone/browser/templates/main_template.pt
 - Location:   (line 94: col 70)
 - Source:     ... 
                                     ^
 - Expression: "context/@@main_template/macros/master"
 - Filename:   ... .8.egg/plone/app/contenttypes/browser/templates/image.pt
 - Location:   (line 6: col 21)
 - Source:     ... tal:use-macro="context/@@main_template/macros/master"
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 - Arguments:  template: <Products.Five.browser.pagetemplatefile.ViewPageTemplateFile object at 0x7efeb0f244c0>
               options: {}
               args: ()
               nothing: None
               modules: <Products.PageTemplates.ZRPythonExpr._SecureModuleImporter object at 0x7efec3b97250>
               request: <WSGIRequest, URL=https://demo-latest-py3.plone.org/en/image/image_view>
               view: <Products.Five.browser.metaconfigure.SimpleViewClass from /home/zope/.cache/eggs/plone.app.contenttypes-2.2.1-py3.8.egg/plone/app/contenttypes/browser/templates/image.pt object at 0x7efeb0f24820>
               context: <Image at /Plone/en/image>
               views: <Products.Five.browser.pagetemplatefile.ViewMapper object at 0x7efeb16504f0>
               here: <Image at /Plone/en/image>
               container: <Image at /Plone/en/image>
               root: <Application at >
               traverse_subpath: []
               user: <PloneUser 'manager'>
               default: <DEFAULT>
               repeat: <Products.PageTemplates.engine.RepeatDictWrapper object at 0x7efeb0fb0e80>
               loop: {}
               target_language: 'en'
               translate: <function BaseTemplate.render.<locals>.translate at 0x7efeb1b9dc10>
               macroname: 'master'
               attrs: {}
```

**Tested with:**
1. Plone 5.2.4.dev0 (5211), CMF 2.4.8, Zope 4.5.3
2. Plone 5.2.3 (5210), CMF 2.4.8, Zope 4.5.3, plone.formwidget.namedfile 2.1.0

I'm not sure if this was the case in older version of Plone. At least I couldn't remember. Also it seems there is a function to handle such cases https://github.com/plone/plone.formwidget.namedfile/blob/1189d0e0349bb5251a26491b4bd6feb096f84a3c/plone/formwidget/namedfile/widget.py#L97-L142

### PR DESCRIPTION
**Pull request by frisi**
this fixes #46

### CODE REVIEW

**Comment by frisi** [comment]
@jensens as you did the last commits/merges on the 2.x branch i assigned you for the review. hope that's okay ;-)

File: news/46.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Uploaded files are set on validation errors [fRiSi]
```
Comment (jensens) [inline news/46.bugfix:1]:
```suggestion
Uploaded files are set on validation errors, see https://github.com/plone/plone.formwidget.namedfile/issues/46  [fRiSi]
```

**Comment by jensens** [comment]
Hey @frisi - it looks good. This needs an PR against master too right?

**Comment by frisi** [comment]
thanks @jensens - please see https://github.com/plone/plone.formwidget.namedfile/pull/51for the PR against master

**Review by thet — APPROVED** [review]
LGTM.

I tested under 5.2 with adding a news item with an image without title (->validation error) and it works.
Same when editing a news items with an image and removing the title.

A little clearer Towncryer entry would be nice. One that says what problem your fix is solving. Like: "Fix issue where already uploaded images were lost when file validation error occurs."

Apparently my original PR only solved the issue that on validation errors previously added images were not lost. Images uploaded during the same request where the validation error occured were lost. This is prevented by this PR.

**Comment by jensens** [comment]
@jenkins-plone-org please run jobs

**Comment by frisi** [comment]
thanks for your review @thet and for amending the comment @jensens 
i'll start jenkins once again and hope the error disappears.

anything else that needs to be done to trigger a release? (eg add in coredev buildout checkouts?)

**Comment by frisi** [comment]
@jenkins-plone-org please run jobs

**Comment by jensens** [comment]
@jenkins-plone-org please run jobs

**Pull request merged**